File: LeetCode/Medium/0198_House_Robber/Solution.java
// 2021-08-27 (tabulation dp)
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        int[] tab = new int[nums.length + 1];
        tab[0] = nums[0];
        tab[1] = nums[1];
        for (int i = 0; i < nums.length; i++) {
            if (tab[i] != 0) {
                if (i + 2 < nums.length) {
                    int skip1 = tab[i] + nums[i + 2];
                    tab[i + 2] = Math.max(tab[i + 2], skip1);
                }
                if (i + 3 < nums.length) {
                    int skip1 = tab[i] + nums[i + 3];
                    tab[i + 3] = Math.max(tab[i + 3], skip1);
                }
                tab[nums.length] = Math.max(tab[i], tab[nums.length]);
            }
        }
        return tab[nums.length];
    }
}

File: LeetCode/Medium/0198_House_Robber/Solution.cpp
// 2022-12-13 (tabulation dp)
class Solution {
public:
    int rob(vector<int>& nums) {
        /*
            input: list of integers with list[i] being $ in the house
            output: maximum amount of money robbable without alerting the police

            notes:
            - cannot rob adjacent houses
            - not circular, so can rob first and last houses

            approach:
            - initialize tabulation dp of max money earnable by that position
            - apply max of current money + money at house[i+2] as well as house[i+3]
            - return the maximum of the list
        */
        int n = nums.size();
        vector<int> maximums = nums;
        for (int i=0; i<n; i++) {
            if (i+2 < n) 
                maximums[i+2] = max(maximums[i+2], maximums[i]+nums[i+2]);
            if (i+3 < n) 
                maximums[i+3] = max(maximums[i+3], maximums[i]+nums[i+3]);
        }

        return *max_element(maximums.begin(), maximums.end());
    }
};